examples/virtual_methods/pet.rs

```diff
@@ -0,0 +1,201 @@
+use glib::prelude::*;
+use glib::subclass::prelude::*;
+
+mod imp {
+    use super::*;
+
+    /// The Pet implementation struct
+    #[derive(Default)]
+    pub struct Pet {}
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for Pet {
+        const NAME: &'static str = "Pet";
+        type Type = super::Pet;
+
+        /// Override the class struct with the custom [`PetClass`]
+        type Class = super::PetClass;
+
+        /// Initialize the class struct with the default implementations of the
+        /// virtual methods.
+        fn class_init(klass: &mut Self::Class) {
+            klass.pet = Some(|obj| obj.imp().pet_default());
+            klass.feed = Some(|obj| obj.imp().feed_default());
+        }
+    }
+
+    impl ObjectImpl for Pet {}
+
+    impl Pet {
+        /// Default implementation of [`PetImpl::pet`]
+        fn pet_default(&self) -> bool {
+            // The default behaviour is unsuccessful pets
+            println!("Pet::pet_default");
+            false
+        }
+
+        /// Default implementation of [`PetImpl::feed`]
+        fn feed_default(&self) {
+            println!("Pet::feed_default");
+        }
+    }
+}
+
+glib::wrapper! {
+    /// The `Pet` class acts as a base class for pets and provides two virtual methods
+    pub struct Pet(ObjectSubclass<imp::Pet>);
+}
+
+/// Custom class struct for the [`Pet`] class.
+///
+/// The class struct is used to implement the vtable for method dispatch. The first field *must*
+/// be a pointer to the parent type. In our case, this is `GTypeInterface`
+#[derive(Copy, Clone, Debug)]
+#[repr(C)]
+pub struct PetClass {
+    /// The first field in a class struct must always be the parent class struct
+    parent_class: glib::gobject_ffi::GObjectClass,
+
+    /// Virtual method for the [`PetImpl::pet`] trait method
+    pub(super) pet: Option<fn(&super::Pet) -> bool>,
+
+    /// Virtual method for the [`PetImpl::feed`] trait method
+    pub(super) feed: Option<fn(&super::Pet)>,
+}
+
+/// Every class struct is required to implement the `ClassStruct` trait
+unsafe impl glib::subclass::types::ClassStruct for PetClass {
+    type Type = imp::Pet;
+}
+
+/// The `PetExt` trait contains public methods of all [`Pet`] objects
+///
+/// These methods need to call the appropriate vfunc from the vtable.
+pub trait PetExt: IsA<Pet> {
+    /// Calls the [`PetImpl::pet`] vfunc
+    fn pet(&self) -> bool {
+        let this = self.upcast_ref();
+        let vtable = this.class();
```

Comment:
```suggestion
        let class = this.class();
```

maybe? Closer to the GObject terminology